### PR TITLE
[JBTM-3148] returning correct error code on one phase commit failure

### PR DESCRIPTION
**Pull request by ochaloup**
https://issues.jboss.org/browse/JBTM-3148

1PC for subordinate XAResource to return heuristics if failure on commit happens.

!QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !RTS  !TOMCAT !JACOCO !LRA !PERF !XTS !AS_TESTS
MAIN

### CODE REVIEW

**Comment by ochaloup** [comment]
@mmusgrov  @tomjenkinson this is my proposal for JBTM-3148. The issue which came from discussion at forum about 1PC XTS (https://developer.jboss.org/thread/279243). The idea of the solution is to report heuristics back to the parent transaction (parent JBoss EAP) as it was implemented by https://issues.jboss.org/browse/JBTM-2443

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java [hunk 1]
```diff
@@ -242,6 +242,22 @@ public int doOnePhaseCommit ()
...
+	        {
+	        // commit was processed correctly but some of the resources failed to finish
+	        // for 2PC this means no trouble as in case of a crash at this time
+	        // there is recovery which works with prepared transaction
+	        // for 1PC where there is nothing prepared and nothing stored at object store
+	        // we need to inform parent about error by setting up heuristic state
+	        case TwoPhaseOutcome.PREPARE_OK:
+	        case TwoPhaseOutcome.FINISH_OK:
+	            if (super.failedList != null && super.failedList.size() > 0) {
+	                return ActionStatus.H_COMMIT;
```
Comment (mmusgrov) [inline ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java]: Is it okay to skip the afterCompletion(H_COMMIT) call that was previously performed at line 252

File: ArjunaJTA/jta/classes/com/arjuna/ats/jta/logging/jtaI18NLogger.java [hunk 2]
```diff
@@ -580,10 +580,14 @@
     @Message(id = 16142, value = "Not actived transaction ''{0}'' for xid: {1}", format = MESSAGE_FORMAT)
     String get_not_activated_transaction(Transaction txn, Xid xid);
 
-    @Message(id = 16143, value = "Problem during waiting for lock '{0}' whilst in state {1}", format = MESSAGE_FORMAT)
+    @Message(id = 16143, value = "Problem during waiting for lock ''{0}'' whilst in state {1}", format = MESSAGE_FORMAT)
```
Comment (mmusgrov) [inline ArjunaJTA/jta/classes/com/arjuna/ats/jta/logging/jtaI18NLogger.java]: Won't this change affect any existing users that are potentially parsing this output? If so you should deprecate first.

Comment (ochaloup) [reply ArjunaJTA/jta/classes/com/arjuna/ats/jta/logging/jtaI18NLogger.java]: This fixes a formatting bug.
The use of `'{0}'` causes that data are not pasted there. Aka the call to method `public void warn_intteruptedExceptionOnWaitingXARecoveryModuleLock(XARecoveryModule recModule, String state, ...` expects to inform user about what recovery module is in use. Now the recovery module is **not** printed. The use of `'{0}'` causes the data is not placed to the output. The change to `''{0}''` makes the information to be printed.

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java (hunk 1 above)
Comment (ochaloup) [reply ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java]: I took the same approach as it's in the `doCommit`: https://github.com/jbosstm/narayana/blob/5.9.5.Final/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java#L170

Plus my understanding is - when the transaction ends with the heuristic outcome we can't say it was completed. Aka [the doc says](https://docs.oracle.com/javaee/7/api/javax/transaction/Synchronization.html#afterCompletion-int-) that the callback is called after transaction is rolled-back or committed. Here we are in *unspecified* state.

**Comment by ochaloup** [comment]
@mmusgrov Thank you for pointing this out. I was wondering about the use of the `TransactionImple.removeTransaction(this);` and I understand it as the transaction is removed from the active transaction to be tracked. It does not mean the participant record is deleted from object store. 

The call is for me: this transaction is not active anymore but reporting back to the caller an error with throwing an exception.
What I was testing the participant record is left in the object store with this code (when the heuristic excepetion is thrown)
Let me know how this could be done better or what are the bad expectations from this perspective. Thanks a lot!

File: ArjunaJTA/jta/classes/com/arjuna/ats/jta/logging/jtaI18NLogger.java (hunk 2 above)
Comment (mmusgrov) [reply ArjunaJTA/jta/classes/com/arjuna/ats/jta/logging/jtaI18NLogger.java]: OK

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java (hunk 1 above)
Comment (mmusgrov) [reply ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java]: The approach taken in `doCommit` for the heuristic outcome is:
https://github.com/jbosstm/narayana/blob/5.9.5.Final/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java#L179
which does call afterCompletion.

**Comment by mmusgrov** [comment]
> @mmusgrov Thank you for pointing this out. I was wondering about the use of the `TransactionImple.removeTransaction(this);` and I understand it as the transaction is removed from the active transaction to be tracked. It does not mean the participant record is deleted from object store.
> 
> The call is for me: this transaction is not active anymore but reporting back to the caller an error with throwing an exception.
> What I was testing the participant record is left in the object store with this code (when the heuristic excepetion is thrown)
> Let me know how this could be done better or what are the bad expectations from this perspective. Thanks a lot!

Okay

**Comment by ochaloup** [comment]
@mmusgrov I found I should add add a specific test for this to testsuite so I'm in middle of work now.

**Comment by ochaloup** [comment]
I worked for tuning end2end test in the QA crashrec testsuite and I added a testcase to the narayana testsuite.

**Comment by ochaloup** [comment]
Pretty strange. There is a failure in the test I added for testing the code changes in this PR. It passes smoothly at my laptop. Investigating...

**Pull request merged**